cmd/access/access_node_builder.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/cmd"
+	"github.com/onflow/flow-go/cmd/build"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/local"
+	"github.com/onflow/flow-go/network"
+	jsoncodec "github.com/onflow/flow-go/network/codec/json"
+	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/network/topology"
+	"github.com/onflow/flow-go/network/validator"
+)
+
+// AccessNodeBuilder is initializes and runs an Access node either as a staked node or an unstaked node
+// It is composed of the FlowNodeBuilder
+type AccessNodeBuilder struct {
+	*cmd.FlowNodeBuilder
+	staked                bool
+	stakedAccessNodeIDHex string
+	unstakedBindAddr      string
+	unstakedNetwork       *p2p.Network
+	unstakedMiddleware    *p2p.Middleware
+}
+
+func FlowAccessNode() *AccessNodeBuilder {
+	return &AccessNodeBuilder{
+		FlowNodeBuilder: cmd.FlowNode(flow.RoleAccess.String()),
+	}
+}
+
+func (anb *AccessNodeBuilder) Initialize() *AccessNodeBuilder {
+
+	anb.PrintBuildVersionDetails()
+
+	anb.BaseFlags()
+
+	anb.ParseAndPrintFlags()
+
+	anb.validateParams()
+
+	// for the staked access node, initialize the network used to communicate with the other staked flow nodes
+	if anb.staked {
+		anb.EnqueueNetworkInit()
+	}
+
+	// if an unstaked bind address is provided, initialize the network to communicate on the unstaked network
+	if anb.unstakedBindAddr != cmd.NotSet {
+		anb.EnqueueUnstakedNetworkInit()
+	}
+
+	anb.EnqueueMetricsServerInit()
+
+	anb.RegisterBadgerMetrics()
+
+	anb.EnqueueTracer()
+
+	return anb
+}
+
+func (anb *AccessNodeBuilder) validateParams() {
+	if anb.staked {
+		return
+	}
+
+	// for an unstaked access node, the staked access node ID must be provided
+	if strings.TrimSpace(anb.stakedAccessNodeIDHex) == "" {
+		anb.Logger.Fatal().Msg("staked access node ID not specified")
+	}
+
+	// and also the unstaked bind address
+	if anb.unstakedBindAddr == cmd.NotSet {
+		anb.Logger.Fatal().Msg("unstaked bind address not set")
+	}
+}
+
+func (anb *AccessNodeBuilder) EnqueueUnstakedNetworkInit() {
+	anb.Component("unstaked network", func(node *cmd.FlowNodeBuilder) (module.ReadyDoneAware, error) {
+		codec := jsoncodec.NewCodec()
```

Comment:
creating the new unstaked network to be used between the unstaked access nodes and the staked access node.